kernel/src/ipc.rs

```diff
@@ -30,22 +29,33 @@ pub enum IPCCallbackType {
 struct IPCData<const NUM_PROCS: usize> {
     /// An array of app slices that this application has shared with other
     /// applications.
-    shared_memory: [Option<AppSlice<SharedReadWrite, u8>>; NUM_PROCS],
+    shared_memory: [ReadWriteAppSlice; NUM_PROCS],
+    search_slice: ReadOnlyAppSlice,
     /// An array of callbacks this process has registered to receive callbacks
     /// from other services.
-    client_callbacks: [Option<Callback>; NUM_PROCS],
+    client_callbacks: [Callback; NUM_PROCS],
     /// The callback setup by a service. Each process can only be one service.
-    callback: Option<Callback>,
+    callback: Callback,
 }
 
 impl<const NUM_PROCS: usize> Default for IPCData<NUM_PROCS> {
     fn default() -> IPCData<NUM_PROCS> {
-        // need this until const_in_array_repeat_expressions is stable
-        const NONE_APPSLICE: Option<AppSlice<SharedReadWrite, u8>> = None;
-        IPCData {
-            shared_memory: [NONE_APPSLICE; NUM_PROCS],
-            client_callbacks: [None; NUM_PROCS],
-            callback: None,
+        unsafe {
+            use core::mem::MaybeUninit;
+            // need this until const_in_array_repeat_expressions is stable
+            let mut array_hack: MaybeUninit<[ReadWriteAppSlice; NUM_PROCS]> = MaybeUninit::uninit();
+            let mut ptr_i = array_hack.as_mut_ptr() as *mut ReadWriteAppSlice;
+            for _ in 0..NUM_PROCS {
+                ptr_i.write(ReadWriteAppSlice::default());
+                ptr_i = ptr_i.add(1);
+            }
+
+            IPCData {
+                shared_memory: array_hack.assume_init(),
+                search_slice: ReadOnlyAppSlice::default(),
+                client_callbacks: [Callback::default(); NUM_PROCS],
+                callback: Callback::default(),
+            }
```

Comment:
Unfortunately, the problem is not that it's not const, it's that `ReadWriteAppSlice` is not marked `Copy`, so without `const_in_array_repeat_expressions` (which is now not even an option with unstable) I don't think there are any tricks we can do.

If we're happy to make `ReadWriteAppSlice` `Copy`, we don't need unsafe, but I believe that could result in mutable aliases.

Reply:
`const_in_array_repeat_expressions` was removed in favor of requiring explicit promotion of constants in constant blocks (see the discussion around [here](https://github.com/rust-lang/rust/issues/49147#issuecomment-736441091)). So as long as there is a const constructor, and you declare the const outside of the actual array instantiation, this works fine. This is the same reason the previous implementation works. Eventually this will get a little cleaner once [inline_const](https://github.com/rust-lang/rust/issues/76001) is stable -- then the `const` could be declared at the point of instantiation.

I tested the code changes above and they compile. Happy to push the commit if you prefer it, or drop it if you don't

Reply:
Aha! I didn't realize the subtlety about declaring the const ouside the array initialization so when I tried it it didn't work. This seems much much much better than what I have, will change.

Reply:
Coincidentally also fixes the imix code size issue